File: examples/ARTestTcpHello/Extension.cpp
#include "TcpVoltageDriver.h"
#include "SetAndMeasureCommand.h"
#if defined(ARTEST_METADATA_GENERATOR)
#include <ARTest/MetadataGenerator.h>
#else
#include <ARTest/Extension.h>
#endif

namespace
{
artest::sdk::Extension DescribeExtension()
{
    using namespace artest::sdk;
    Extension extension{"com.artest.example.tcp-hello", "0.1.0", "TCP Hello World", "ARTest"};

    extension.AddDriver<artest::tcphello::TcpVoltageDriver>({
        .id = "com.artest.example.driver.tcp-voltage", .name = "Loopback voltage driver",
        .contract = artest::tcphello::Contract, .mode = DriverMode::Simulated,
        .metadata = {.schema = Schema::Object()
            .Required("port", Schema::Integer().Minimum(1).Maximum(65535))
            .Optional("initializeMs", Schema::Integer().Minimum(1).Maximum(10000))
            .Optional("operationMs", Schema::Integer().Minimum(1).Maximum(10000))
            .Optional("shutdownMs", Schema::Integer().Minimum(1).Maximum(10000)),
            .schemaId = "artest.schema.example.tcp-voltage.configuration.v1"}});

    extension.AddCommand<artest::tcphello::SetAndMeasureCommand>({
        .id = "com.artest.example.command.tcp-set-and-measure", .name = "Set and measure voltage",
        .metadata = {.schema = Schema::Object()
            .Required("voltage", Schema::Number().Minimum(0).Maximum(60))
            .Required("minimum", Schema::Number().Minimum(0).Maximum(60))
            .Required("maximum", Schema::Number().Minimum(0).Maximum(60)),
            .schemaId = "artest.schema.example.tcp-measurement.parameters.v1",
            .requiredContracts = {artest::tcphello::Contract}}});

    return extension;
}
}
#if defined(ARTEST_METADATA_GENERATOR)
ARTEST_GENERATE_METADATA(DescribeExtension)
#else
ARTEST_EXPORT_EXTENSION(DescribeExtension)
#endif

File: examples/ARTestTcpHello/python/extension.py
"""Same command semantics as C++; transport belongs exclusively to the native driver."""
from dataclasses import dataclass
from artest_sdk import Command, Extension, Result, parameter

CONTRACT = "artest.contract.example.tcp-voltage.v1"

@dataclass
class Measurement:
    voltage: float = parameter(default=12.0, minimum=0, maximum=60)
    minimum: float = parameter(default=11.9, minimum=0, maximum=60)
    maximum: float = parameter(default=12.1, minimum=0, maximum=60)

class SetAndMeasure(Command):
    async def execute(self, parameters, context):
        if parameters.minimum > parameters.maximum:
            return Result(status="invalidArgument", message="minimum must not exceed maximum")
        async with context.instrument(CONTRACT) as instrument:
            result = await instrument.invoke("artest.example.tcp-voltage.v1/apply",
                                             {"voltage": parameters.voltage})
        value = result.data["value"]
        return Result.verdict(parameters.minimum <= value <= parameters.maximum,
            {"value": value, "unit": "V", "minimum": parameters.minimum,
             "maximum": parameters.maximum, "instrumentId": context.instrument_id},
            "artest.schema.example.tcp-measurement.v1", "TCP voltage limits")

def define_extension():
    extension = Extension("com.artest.example.python.tcp-hello", "0.1.0", "Python TCP Hello command", "ARTest")
    extension.command("com.artest.example.python.command.tcp-set-and-measure", SetAndMeasure,
                      Measurement, name="Set and measure voltage (Python)", requires=(CONTRACT,))
    return extension
